earth2studio/data/mrms.py

```diff
@@ -0,0 +1,427 @@
+        -------
+        xr.DataArray
+                Data array for the specified time with dimensions ``[time, variable, y, x]``.
+
+        Raises
+        ------
+        FileNotFoundError
+                If the requested remote file does not exist.
+        KeyError
+                If a requested variable is not present in the MRMS lexicon.
+        """
+        if time.tzinfo is None:
+            time = time.replace(tzinfo=timezone.utc)
+
+        # Resolve to an exact available S3 object within tolerance
+        resolved = self._resolve_s3_time(time)
+        if resolved is None:
+            raise FileNotFoundError(
+                f"No MRMS file found near requested time {time.isoformat()} "
+                f"within ±{self.max_offset_minutes} minutes for product {self.product}"
+            )
+        actual_time, s3_uri = resolved
+
+        if self._verbose:
+            logger.info(f"Fetching MRMS file: {s3_uri}")
+
+        grib_path = self._download_and_decompress(s3_uri)
+
+        # Load with cfgrib.
+        ds = xr.open_dataset(
+            grib_path, engine="cfgrib", backend_kwargs={"indexpath": ""}
+        )
+        if len(ds.data_vars) == 0:
+            raise RuntimeError(f"No data variables found in MRMS file: {s3_uri}")
+        data_var_name = list(ds.data_vars)[0]
+        field = ds[data_var_name].rename({"latitude": "lat", "longitude": "lon"})
+
+        # Build output DataArray with [time, variable, lat, lon]
+        # Assumes lat and lon are 1D in source grib2 files
+        lat_dim, lon_dim = field.dims[-2], field.dims[-1]
+        lat_size, lon_size = field.sizes[lat_dim], field.sizes[lon_dim]
+
+        out = xr.DataArray(
+            data=np.empty((1, len(variables), lat_size, lon_size), dtype=field.dtype),
+            dims=["time", "variable", "lat", "lon"],
+            coords={
+                "time": [actual_time],
+                "variable": variables,
+                "lat": field.coords["lat"],
+                "lon": field.coords["lon"],
+            },
+        )
+
+        # Map each requested variable via MRMS lexicon; MRMS data contains the product already
+        for i, v in enumerate(variables):
+            # Translate to ensure variable is valid for MRMS
+            try:
+                _, modifier = MRMSLexicon[v]
+            except KeyError as e:
+                raise KeyError(f"Variable id {v} not found in MRMS lexicon.") from e
+            out[0, i] = modifier(field.values)
+
+        return out
+
+    def _download_and_decompress(self, s3_uri: str) -> str:
+        """Download gzipped GRIB2 from S3 and decompress into cache; return path."""
+        # Cache filenames derived from key
+        key_hash = (
+            s3_uri.replace("s3://", "").replace("/", "_").replace(".grib2.gz", "")
+        )
+        gz_path = os.path.join(self.cache, f"{key_hash}.grib2.gz")
+        grib_path = os.path.join(self.cache, f"{key_hash}.grib2")
+
+        # Download gz if not present
```

Comment:
Why keep both? Could you just check to see if the .grib2 is present and if not you download and decompress?